src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -0,0 +1,842 @@
+        /// </summary>
+        /// <exception cref="JsonWriterException">
+        /// Thrown if this would result in an invalid JSON to be written (while validation is enabled).
+        /// </exception>
+        public void WriteEndObject()
+        {
+            WriteEnd(JsonConstants.CloseBrace);
+            _tokenType = JsonTokenType.EndObject;
+        }
+
+        private void WriteEnd(byte token)
+        {
+            _currentDepth |= 1 << 31;
+            _currentDepth--;
+
+            // Necessary if WriteEndX is called without a corresponding WriteStartX first.
+            // Checking for int.MaxValue because int.MinValue - 1 = int.MaxValue
+            if (_currentDepth == int.MaxValue)
+            {
+                _currentDepth = 0;
+            }
+
+            if (_writerOptions.SlowPath)
+                WriteEndSlow(token);
+            else
+                WriteEndMinimized(token);
+        }
+
+        private void WriteEndMinimized(byte token)
+        {
+            while (_buffer.Length < 1)
+            {
+                GrowAndEnsure();
+            }
+
+            _buffer[0] = token;
+            Advance(1);
+        }
+
+        private void WriteEndSlow(byte token)
+        {
+            Debug.Assert(_writerOptions.Indented || !_writerOptions.SkipValidation);
+
+            if (_writerOptions.Indented)
+            {
+                if (!_writerOptions.SkipValidation)
+                {
+                    ValidateEnd(token);
+                }
+                WriteEndIndented(token);
+            }
+            else
+            {
+                Debug.Assert(!_writerOptions.SkipValidation);
+                ValidateEnd(token);
+                WriteEndMinimized(token);
+            }
+        }
+
+        private void ValidateEnd(byte token)
+        {
+            if (_bitStack.CurrentDepth <= 0)
+                ThrowHelper.ThrowJsonWriterException(ExceptionResource.MismatchedObjectArray, token);
+
+            if (token == JsonConstants.CloseBracket)
+            {
+                if (_inObject)
+                {
+                    Debug.Assert(_tokenType != JsonTokenType.None);
+                    ThrowHelper.ThrowJsonWriterException(ExceptionResource.MismatchedObjectArray, token);
+                }
+            }
+            else
+            {
+                Debug.Assert(token == JsonConstants.CloseBrace);
+
+                if (!_inObject)
+                {
+                    ThrowHelper.ThrowJsonWriterException(ExceptionResource.MismatchedObjectArray, token);
+                }
+            }
+
+            _inObject = _bitStack.Pop();
+        }
+
+        private void WriteEndIndented(byte token)
+        {
+            // Do not format/indent empty JSON object/array.
+            if ((_tokenType == JsonTokenType.StartObject && token == JsonConstants.CloseBrace)
+                || (_tokenType == JsonTokenType.StartArray && token == JsonConstants.CloseBracket))
```

Comment:
The second half of the ands should always be true, right?  You could just assert that the previous methods matched appropriately... seems weird that if it gets misaligned you'll print a newline.

Reply:
Agreed. Will fix.